src/artifacts.js

```diff
@@ -254,23 +325,85 @@ api.declare({
         });
     }
 
+    if (artifact.storageType === 'blob') {
+      if (artifact.details.uploadId) {
+        // TODO: Here's where we return a new set of signed requests to do
+        // with this.s3Controller.genereate....();
+      }
+    }
+
     // Update expiration and detail, which may have been modified
     await artifact.modify((artifact) => {
       artifact.expires = expires;
       artifact.details = details;
     });
   }
 
-  // Publish message about artifact creation
-  await this.publisher.artifactCreated({
-    status:         task.status(),
-    artifact:       artifact.json(),
-    workerGroup,
-    workerId,
-    runId,
-  }, task.routes);
+
+  // This event is *invalid* for s3/azure storage types so we'll stop sending it.
+  // It's only valid for error, reference and blob
+  if (artifact.storageType === 'error' || artifact.storageType === 'reference') {
+    // Publish message about artifact creation
+    await this.publisher.artifactCreated({
+      status:         task.status(),
+      artifact:       artifact.json(),
+      workerGroup,
+      workerId,
+      runId,
+    }, task.routes);
+  }
 
   switch (artifact.storageType) {
+    case 'blob':
+      let requests;
+      let uploadId;
+      // If we're supposed to do a multipart upload, we should generate an UploadId
+      // if it doesn't already exist.  We should store that ID in the entity
+      if (artifact.details.parts) {
+        if (!artifact.details.uploadId) {
+          uploadId = await this.s3Controller.initiateMultipartUpload({
+            bucket: artifact.details.bucket,
+            key: artifact.details.key,
+            sha256: artifact.details.sha256,
+            size: artifact.details.size,
+            metadata: {taskId, runId, name},
+            tags: {taskId, runId, name},
+            contentType: artifact.contentType,
+            contentEncoding: artifact.details.encoding || undefined,
+          });
+          await artifact.modify(_artifact => {
+            _artifact.details.uploadId = uploadId;
```

Comment:
We should do:
```js
await artifact.modify(artifact => {
  if (!artifact.details.uploadId) {
    artifact.details.uploadId = uploadId;
  }
});
uploadId = artifact.details.uploadId;
```

That way if we're racing it's always the first one that wins...

We could add:
```js
if (artifact.details.uploadId != upload) {
  await s3.cancelMultiPartUpload(uploadId);
  uploadId = artifact.details.uploadId;
}
```
But the cancel part is just nice...

Note: I think all of this should be part of the create entity phase... and that we then handle this cancel uploadId part if there is a create conflict...